File: 2/PDOO/NapakalakiJava/src/napakalakiGame/Player.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package napakalakiGame;

import GUI.Dice;
import java.util.ArrayList;
import java.util.Random;

/**
 *
 * @author david
 */
public class Player {
    private String name;
    private static final int MAXLEVEL=10;
    private int level=1;
    private boolean dead=true;
    private boolean canISteal=true;
    protected Player enemy;
    private ArrayList<Treasure> visibleTreasures=new ArrayList();
    private ArrayList<Treasure> hiddenTreasures=new ArrayList();
    private BadConsequence pendingBadConsequence=null;
    
    public Player(String name){
       this.name=name;
    }
    
    public static int getMAXLEVEL(){
        return MAXLEVEL;
    }
    
    public Player(Player p){
        name=p.name;
        level=p.level;
        dead=p.dead;
        canISteal=p.canISteal;
        enemy=p.enemy;
        visibleTreasures=new ArrayList(p.visibleTreasures);
        hiddenTreasures=new ArrayList(p.hiddenTreasures);
        pendingBadConsequence=p.pendingBadConsequence;
    }
    
    public String getName(){
        return this.name;
    }
    
    private void bringToLife(){
        this.dead=false;
    }
    
    protected int getCombatLevel(){
        int l=level;
        for (Treasure t: visibleTreasures){
            l+=t.getBonus();
        }
        return l;
    }
    
    private void incrementLevels(int l){
        if (this.level+l< MAXLEVEL){
            level+=l;
        }
        else {
            level= MAXLEVEL;
        }
        
    }
    
    private void decrementLevels(int l){
        if (this.level-l> 0){
            level-=l;
        }
        else {
            level=1;
        }
    }
    
    private void setPendingBadConsequence(BadConsequence b){
        pendingBadConsequence=b;
    }
    
    
    private void applyPrize(Monster m){
        int nLevels=m.getLevelsGained();
        incrementLevels(nLevels);
        int nTreasures=m.getTreasuresGained();
        if (nTreasures>0){
            CardDealer dealer=CardDealer.getInstance();
            for (int i=0; i<nTreasures; ++i){
                Treasure t=dealer.nextTreasure();
                hiddenTreasures.add(t);
            }
        }
    }
    
    private void applyBadConsequence(Monster m){
        BadConsequence badConsequence=m.getBc();
        int nLevels=badConsequence.getLevels();
        decrementLevels(nLevels);
        BadConsequence pendingBad=badConsequence.adjustToFitTreasureLists(visibleTreasures, hiddenTreasures);        
        setPendingBadConsequence(pendingBad);
    }
    
    private boolean canMakeTreasureVisible(Treasure t){
        int contador=0;
        if (pendingBadConsequence!=null && !pendingBadConsequence.isEmpty())
            return false;
        if(t.getType()==TreasureKind.BOTHHANDS){
             for(Treasure t1: visibleTreasures)
                 if (t1.getType()==TreasureKind.BOTHHANDS || t1.getType()==TreasureKind.ONEHAND)
                     return false;
             
        }
        else if (t.getType()==TreasureKind.ONEHAND){
            for (Treasure t1: visibleTreasures)
                if (t1.getType()==TreasureKind.ONEHAND)
                    contador++;
                else if (t1.getType()==TreasureKind.BOTHHANDS)
                    return false;
            return contador<2;
        }
        else
            for (Treasure t1: visibleTreasures)
                if (t1.getType()==t.getType())
                    return false;
        return true;
    }
    
    private int howManyVisibleTreasures(TreasureKind tKind){
        int contador=0;
        for (Treasure t: visibleTreasures){
            if (t.getType()==tKind){
                if (tKind==TreasureKind.BOTHHANDS){
                    contador+=2;
            }
                else {
                    contador++;
                }
        }
        }
        
        return contador;
    }
    
    private void dieifNoTreasures(){
        if (hiddenTreasures.isEmpty() && visibleTreasures.isEmpty())
            dead=true;
    }
    
    public boolean isDead(){
        return dead;
    }
    
    public ArrayList<Treasure> getHiddenTreasures(){
        return hiddenTreasures;
    }
    
    public ArrayList<Treasure> getVisibleTreasures(){
        return visibleTreasures;
    }
    
    public CombatResult combat(Monster m){
        int myLevel=getCombatLevel();
        int monsterLevel=this.getOponentLevel(m);
        if (myLevel>monsterLevel){
            applyPrize(m);
            if (level>=MAXLEVEL)
                return CombatResult.WINGAME;
            else
                return CombatResult.WIN;
        }
        else{
            applyBadConsequence(m);
            if(shouldConvert())
                return CombatResult.LOSEANDCONVERT;
            else
                return CombatResult.LOSE;
        }
        
    }
    
    public void makeTreasureVisible(Treasure t){
        boolean canI=canMakeTreasureVisible(t);
        if (canI){
            visibleTreasures.add(t);
            hiddenTreasures.remove(t);
        }
    }
    
    public void discardVisibleTreasure(Treasure t){
        visibleTreasures.remove(t);
        if (pendingBadConsequence!=null && !pendingBadConsequence.isEmpty())
            pendingBadConsequence.substractVisibleTreasure(t);
        dieifNoTreasures();
    }
    
    public void discardHiddenTreasure(Treasure t){
        hiddenTreasures.remove(t);
        if (pendingBadConsequence!=null && !pendingBadConsequence.isEmpty())
            pendingBadConsequence.substractHiddenTreasure(t);
        dieifNoTreasures();

    }
    
    public boolean validState(){
        return (pendingBadConsequence==null || pendingBadConsequence.isEmpty())
                && hiddenTreasures.size()<=4;
    }
        
    
    public void initTreasures(){
        CardDealer dealer=CardDealer.getInstance();
        Dice dice=Dice.getInstance();
        bringToLife();
        hiddenTreasures.add(dealer.nextTreasure());
        int number=dice.nextNumber();
        if (number>1)
            hiddenTreasures.add(dealer.nextTreasure());
        if (number==6)
            hiddenTreasures.add(dealer.nextTreasure());
    }
    
    public int getLevels(){
        return level;
    }
    
    public Treasure stealTreasure(){
        Treasure treasure=null;
        boolean canI=canISteal();
        if (canI){
            boolean canYou = enemy.canYouGiveMeATreasure();
            if (canYou){
                treasure=enemy.giveMeATreasure();
                hiddenTreasures.add(treasure);
                haveStolen();
            }
        }
        return treasure;
    }
    
    public Player getEnemy() {
      return enemy;
    }
    
    public void setEnemy(Player enemy){
        this.enemy=enemy;
        
    }
    public BadConsequence getPendingBadConsequence() {
      return pendingBadConsequence;
    }
    protected Treasure giveMeATreasure(){
        Random rnd=new Random();
        int a=(int)(rnd.nextDouble() * hiddenTreasures.size());
        return hiddenTreasures.remove(a);
    }
    
    public boolean canISteal(){
        return canISteal;
    }
    
    public boolean canYouGiveMeATreasure(){
        return !hiddenTreasures.isEmpty();
    }
    
    private void haveStolen(){
        canISteal=false;
        
    }
    
    public void discardAllTreasures(){
        ArrayList<Treasure> copia=new ArrayList(visibleTreasures);
        for (Treasure t: copia)
            discardVisibleTreasure(t);
        
        copia=new ArrayList(hiddenTreasures);
        for (Treasure t: copia)
            discardHiddenTreasure(t);
    }
    public String toString(){
        String a="Nombre: " + name + "\n" +
        " Nivel: " + Integer.toString(level) + "\n" +
        " Enemigo: " + enemy.getName() + "\n";
        return a;
                    
    }
    
    protected int getOponentLevel(Monster m){
        return m.getLevel();
    }
    
    protected boolean shouldConvert(){
        Dice dice=Dice.getInstance();
        int n =dice.nextNumber();
        if(n==1){
            return true;
        }
        else{
            return false;
        }
    }
}

 

File: 2/PDOO/NapakalakiRuby/lib/player.rb

#encoding: utf-8
 require_relative 'bad_consequence'
 require_relative 'combat_result'
 require_relative 'dice'
 require_relative 'treasure'
 require_relative 'treasure_kind'
 module NapakalakiGame
class Player
  attr_reader :name, :dead, :level, :canISteal, :hiddenTreasures, :visibleTreasures, :enemy
  attr_reader :pendingBadConsequence
  attr_writer :pendingBadConsequence, :enemy
  
  @@MAXLEVEL=10
  def initialize(name)
    @enemy
    @name=name
    @level=1
    @dead=true
    @canISteal=true
    @hiddenTreasures=Array.new
    @visibleTreasures=Array.new
    @pendingBadConsequence
    
  end
  
  def bringToLife
    @dead=false;
  end
  
  def copy(p)
    @enemy=p.enemy
    @name=p.name
    @level=p.level
    @dead=p.dead
    @canISteal=p.canISteal
    @hiddenTreasures=Array.new(p.hiddenTreasures)
    @visibleTreasures=Array.new(p.visibleTreasures)
    @pendingBadConsequence=p.pendingBadConsequence
    puts "ERES CULTIST!!!!!!!!!!!!!!"
  end
  def self.MAXLEVEL
    @@MAXLEVEL
  end
  
  def incrementLevels(l)
    if ((@level+l)<@@MAXLEVEL)
      @level=@level+l
    else
      @level=@@MAXLEVEL
    end
  end
  
  def decrementLevels(l)
    if ((@level-l)>0)
      @level=@level-l
    else
      @level=1
    end
  end
  
  def applyPrize(m)
    nLevels=m.price.level
    incrementLevels(nLevels)
    nTreasures=m.price.treasures
    if (nTreasures>0)
      dealer=CardDealer.instance
      for i in 0..nTreasures-1 do
        t=dealer.nextTreasure
        @hiddenTreasures << t
      end
    end
  end
  
  def applyBadConsequence(m)
    badConsequence=m.bc
    nLevels=badConsequence.levels
    decrementLevels(nLevels);
    @pendingBadConsequence=badConsequence.adjustToFitTreasureList(@visibleTreasures, @hiddenTreasures);        
    
  end
  
  def canMakeTreasureVisible(t)
    return false if (!@pendingBadConsequence.nil? && !@pendingBadConsequence.isEmpty)
    if (t.type==TreasureKind::BOTHHANDS)
      @visibleTreasures.each {|t1| return false if (t1.type==TreasureKind::BOTHHANDS || t1.type==TreasureKind::ONEHAND)}
    elsif (t.type==TreasureKind::ONEHAND)
      contador=0;
      @visibleTreasures.each do |t1|
          contador += 2 if t1.type==TreasureKind::BOTHHANDS
          contador += 1 if t1.type==TreasureKind::ONEHAND
      end
      return contador<2
    else
      @visibleTreasures.each{|t1| return false if t.type==t1.type}
    end
    true
      
  end
    
 
  
  def howManyVisibleTreasures(tKind)
    contador=0
    @visibleTreasures.each do |t|
      if (t.type==tKind)
          if (tKind==TreasureKind::BOTHHANDS)
            contador=contador+2
          else
            contador=contador+1
          end
      end
    end
    contador
  end
  
  def dieIfNoTreasures
    if (@hiddenTreasures.empty? && @visibleTreasures.empty?)
      @dead=true
    end
  end
  
  def combat(m)
    myLevel=getCombatLevel
    monsterLevel=getOponentLevel(m)
        if (myLevel>monsterLevel)
            applyPrize(m)
            if (level>=@@MAXLEVEL)
                CombatResult::WINGAME
            else
                CombatResult::WIN
            end
        
        else
            applyBadConsequence(m)
            if(shouldConvert)
              CombatResult::LOSEANDCONVERT
            else
              CombatResult::LOSE
            end
        end
  end
  
  def makeTreasureVisible(t)
    canI=canMakeTreasureVisible(t);
        if (canI)
            @visibleTreasures << (t);
            @hiddenTreasures.delete(t);
        end
  end
  
  def discardVisibleTreasure(t)
    @visibleTreasures.delete(t)
        if (!@pendingBadConsequence.nil? && !@pendingBadConsequence.isEmpty)
            @pendingBadConsequence.substractVisibleTreasure(t);
        end
        dieIfNoTreasures 
  end
  
  def discardHiddenTreasure(t)
    @hiddenTreasures.delete(t)
        if (!@pendingBadConsequence.nil? && !@pendingBadConsequence.isEmpty)
            @pendingBadConsequence.substractHiddenTreasure(t);
        end
        dieIfNoTreasures
  end
  
  def validState
    (@pendingBadConsequence.nil? || @pendingBadConsequence.isEmpty) && 
    @hiddenTreasures.length<=4
  end
  
  def initTreasures
    dealer=CardDealer.instance
    dice=Dice.instance
    bringToLife
    @hiddenTreasures << dealer.nextTreasure
    number=dice.nextNumber
    @hiddenTreasures << dealer.nextTreasure if number>1
    @hiddenTreasures << dealer.nextTreasure if number==6
  end
  
  def stealTreasure
        canI=canISteal();
        if (canI)
            canYou = @enemy.canYouGiveMeATreasure
            if (canYou)
                treasure=@enemy.giveMeATreasure
                @hiddenTreasures << treasure
                haveStolen
            end
        end
        treasure;
  end
  
  def giveMeATreasure
    @hiddenTreasures.delete_at(rand(@hiddenTreasures.size))
  end
  
  def canYouGiveMeATreasure
    !@hiddenTreasures.empty?
  end
  def getCombatLevel
    l=@level;
    @visibleTreasures.each{|t| l=l+t.bonus}
    l
  end
  def haveStolen
    @canISteal=false
  end
  
  def discardAllTreasures
     for i in 0..@visibleTreasures.size-1 do
       discardVisibleTreasure(@visibleTreasures[0])
     end
     for i in 0..@hiddenTreasures.size-1 do
       discardHiddenTreasure(@hiddenTreasures[0])
     end
  end
  
  def getVisibleTreasures
    @visibleTreasures
  end
  
  def getHiddenTreasures
    @hiddenTreasures
  end
  
  def getName
    @name
  end
  
  def to_s
    "#{@name}\n" +
    "Nivel: #{@level}\n" +
    "Enemigo:#{@enemy.name}\n" +
    "Mala cons:\n#{@pendingBadConsequence}\n"
  end
  
  def getOponentLevel(m)
    m.combatLevel
  end
  
  def shouldConvert
    dice=Dice.instance
    number=dice.nextNumber
    number==1
  end
  private :bringToLife, :incrementLevels, :decrementLevels, :applyPrize, :applyBadConsequence
  private :canMakeTreasureVisible, :howManyVisibleTreasures, :dieIfNoTreasures, :haveStolen
  private :getCombatLevel
  protected :canYouGiveMeATreasure, :giveMeATreasure, :getCombatLevel, :getOponentLevel
  protected :shouldConvert
end
 end
